include/nbl/builtin/hlsl/complex.hlsl

```diff
@@ -0,0 +1,191 @@
+        {
+                complex_t retVal = { real, imag };
+                return retVal;
+        }
+            
+        static complex_t create(const complex_t other)
+        {
+                return other;
+        }
+            
+        // ------------------------- Member functions -------------------------------      
+        Scalar real() {
+            return m_real;
+        }
+            
+        void real(const Scalar value)
+        {
+            m_real = value;
+        }
+            
+        Scalar imag()
+        {
+            return m_imag;
+        }
+            
+        void imag(const Scalar value)
+        {
+            m_imag = value;
+        }
+        // ------------------------- Arithmetic operators -------------------------------   
+        complex_t operator+(const complex_t rhs)
+        {
+            complex_t result;
+            result.m_real = m_real + rhs.m_real;
+            result.m_imag = m_imag + rhs.m_imag;
+
+            return result;
+        }
+            
+        complex_t operator-(const complex_t rhs)
+        {
+            complex_t result;
+
+            result.m_real = m_real - rhs.m_real;
+            result.m_imag = m_imag - rhs.m_imag;
+
+            return result;
+        }
+
+        complex_t operator*(const complex_t rhs)
+        {
+            complex_t result;
+
+            result.m_real = m_real * rhs.m_real - m_imag * rhs.m_imag;
+            result.m_imag = m_real * rhs.m_imag + m_imag * rhs.m_real;
+
+            return result;
+        }
+            
+        // multiply by scalar
+        complex_t operator*(const Scalar scalar)
+        {
+            complex_t result;
+            result.m_real = m_real * scalar;
+            result.m_imag = m_imag * scalar;
+
+            return result;
+        }
+            
+        // Divide by scalar
+        complex_t operator/(const Scalar scalar)
+        {
+            complex_t result;
+            result.m_real = m_real / scalar;
+            result.m_imag = m_imag / scalar;
+
+            return result;
+        }
+            
+        complex_t operator/(const complex_t rhs)
+        {
+            complex_t result;
+
+            T denominator = rhs.m_real * rhs.m_real + rhs.m_imag * rhs.m_imag;
+            result.m_real = (m_real * rhs.m_real + m_imag * rhs.m_imag) / denominator;
+            result.m_imag = (m_imag * rhs.m_real - m_real * rhs.m_imag) / denominator;
+
+            return result;
+        }
+            
+        // ----------------- Relational operators -----------------------------
+        bool operator==(const complex_t rhs)
+        {
+                return !(uint64_t(m_real) ^ uint64_t(rhs.m_real)) && !(uint64_t(m_imag) ^ uint64_t(rhs.m_imag));
+        }
+        bool operator!=(const complex_t rhs)
+        {
+                return uint64_t(m_real) ^ uint64_t(rhs.m_real) || uint64_t(m_imag) ^ uint64_t(rhs.m_imag);
```

Comment:
don't cast to `uint64_t` :
1. you're not comparing bitpatterns, you're casting to uint64_T which truncatest the floats, etc.
2. `Scalar=float32_t` would have been 32bit
3. `!=` and `==` work on floats, so use them